### PR TITLE
[Api][Checkout] Allowing access only for correctly logged in

### PR DESCRIPTION
**Pull request by Tomanhez**
| Q               | A
| --------------- | -----
| Branch?         | 1.8
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

### CODE REVIEW

File: features/cart/shopping_cart/allowing_access_only_for_correctly_logged_in.feature [hunk 1]
```diff
@@ -0,0 +1,45 @@
+@accessing_cart
+Feature: Allowing access only for correctly logged in
```
Comment (GSadee) [inline features/cart/shopping_cart/allowing_access_only_for_correctly_logged_in.feature:2]:
```suggestion
Feature: Allowing access only for correctly logged in users
```

File: features/cart/shopping_cart/allowing_access_only_for_correctly_logged_in.feature [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    I want to be able to have access to cart for users with properly access
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And the store has a product "Stark T-shirt" priced at "$12.00"
+
+    @api
+    Scenario: Visitor has access to his cart
+        When there is the visitor
+        Then the visitor can add "Stark T-Shirt" product in the cart
```
Comment (GSadee) [inline features/cart/shopping_cart/allowing_access_only_for_correctly_logged_in.feature:14]: Maybe sth more like:
```suggestion
    Scenario: Accessing to the cart by the visitor
        When the visitor adds "Stark T-Shirt" product to the cart
        Then the visitor can see "Stark T-Shirt" product in the cart
```

**Comment by GSadee** [comment]
Thank you, Tomasz! :1st_place_medal:

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 3]
```diff
@@ -74,6 +74,15 @@ public function iSeeTheSummaryOfMyCart(string $tokenValue): void
         $this->cartsClient->show($tokenValue);
     }
 
+    /**
+     * @When /^the (visitor|administrator) try to see the summary of (customers|visitor) (cart)$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:78]: `(customers|visitor)` -> `(customer|visitor)'s`

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 4]
```diff
@@ -74,6 +74,15 @@ public function iSeeTheSummaryOfMyCart(string $tokenValue): void
...
+     * @When /^the (visitor|customer) see the summary of (his) (cart)$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:79]: `his` -> `their`

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 5]
```diff
@@ -74,6 +74,15 @@ public function iSeeTheSummaryOfMyCart(string $tokenValue): void
...
+     */
+    public function theSeeTheSummaryOfCart(string $userType, string $userType2, string $tokenValue): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:81]: `string $userType, string $userType2` could be removed

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 6]
```diff
@@ -110,6 +119,19 @@ public function iChangeQuantityToInMyCart(ProductInterface $product, int $quanti
         $this->changeQuantityOfOrderItem($itemId, $quantity, $tokenValue);
     }
 
+    /**
+     * @When /^the (visitor|customer) change (product "[^"]+") quantity to (\d+) in his (cart)$/
+     * @When /^the (visitor) try to change (product "[^"]+") quantity to (\d+) in the customer (cart)$/
+     */
+    public function theVisitorChangeProductQuantityToInHisCart(
+        string $userType,
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:127]: Should be removed as you do not use it

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 7]
```diff
@@ -145,6 +167,7 @@ public function myCartSTotalShouldBe(string $tokenValue, int $total): void
 
     /**
      * @Then /^my (cart) should be empty$/
+     * @Then /^the visitor has no access to customers (cart)$/
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:170]: `customer's`

File: src/Sylius/Behat/Context/Api/Shop/CartContext.php [hunk 8]
```diff
@@ -228,11 +251,21 @@ public function thisItemShouldHaveCode(array $item, string $variantCode): void
         );
     }
 
+    /**
+     * @Then /^the administrator should see ("[^"]+" product) with quantity ([^"]+) in the (customer|visitor) cart$/
+     */
+    public function theAdministratorShouldSeeProductWithQuantityInTheCart(
+        string $productName,
+        int $quantity,
+        string $userType
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/CartContext.php:260]: Remove every user type that is no used

File: src/Sylius/Behat/Context/Api/Shop/LoginContext.php [hunk 9]
```diff
@@ -27,6 +27,14 @@ public function __construct(ApiSecurityClientInterface $client)
         $this->client = $client;
     }
 
+    /**
+     * @Given there is the visitor
+     */
+    public function iAmAVisitor(): void
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/LoginContext.php:33]: Is it used somewhere?

Comment (Tomanhez) [reply src/Sylius/Behat/Context/Api/Shop/LoginContext.php:33]: https://github.com/Sylius/Sylius/pull/11806/files#diff-6e397cfc7a6739ac80d9d7bc67120c01R191

File: src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderPatchMethodItemExtension.php [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    public function applyToItem(
+        QueryBuilder $queryBuilder,
+        QueryNameGeneratorInterface $queryNameGenerator,
+        string $resourceClass,
+        array $identifiers,
+        string $operationName = null,
+        array $context = []
+    ) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ApiBundle/Doctrine/QueryItemExtension/OrderPatchMethodItemExtension.php:45]: missing return type

File: src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php [hunk 11]
```diff
@@ -35,6 +35,12 @@
 
 final class CheckoutContext implements Context
 {
+    const CHECKOUT_STATE_TYPES = [
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Api/Shop/CheckoutContext.php:38]:
```suggestion
    private const CHECKOUT_STATE_TYPES = [
```

**Pull request merged**